gamla/functional.py

```diff
@@ -436,3 +436,19 @@ def countby_many(f, it):
         curried.map(f),
         groupby_many_reduce(toolz.identity, lambda x, y: x + 1 if x else 1),
     )
+
+
+@currying.curry
+def take_while(pred, seq):
+    for x in seq:
+        if not pred(x):
+            return
+        yield x
+
+
+@currying.curry
+def take_last_while(pred, seq):
```

Comment:
if you implement as a reducer, you can avoid keeping everything in memory for cases where the tail is short.

the reduce will go over the iterable, if an `element` is falsy on `pred`, reset `state`, otherwise keep appending to `state` on the left. at the end just return it. make sure you append in constant time.